[Workflow] Add PlantUML dumper to workflow:dump command

Plopix opened the pull request:
| Q             | A
| ------------- | ---
| Branch        | 4.1
| Bug fix       | no
| New feature   | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/9220

Hello,

`workflow:dump` exports workflows in `dot` format. To me, this format is not the easiest and not the simplest to use. Styles and logic are mixed together which makes it hard to read.

[PlantUML](http://plantuml.com/) is a tool based on Graphviz like Dot but that 
generates nicer diagrams, more readable and most of all easier to adapt to your
documentation. Just copy and paste the PUML on the website you will see the rendering live.

Also, there is a [PHPStorm Plugin](https://plugins.jetbrains.com/plugin/7017-plantuml-integration) and [plenty of integration](http://plantuml.com/running) of this format.

This PR adds 2 options

* a `--dump-format=puml` option to the `workflow:dump` command to generate the workflows in PlantUML.

The conversion requires the PlantUML JAR, and can be used like that:
```bash
php bin/console workflow:dump pull_request --dump-format=puml | java -jar plantuml.jar -p  > workflow.png
```
> don't forget the `-p` to enable the "piping"

Here is an example with `pull_request` workflow of the documentation (with no style and no marking):
```
@startuml
title pull_request
state start <<initial>>
state coding
state travis
state review
state merged
state closed
start --> travis: submit
coding --> travis: update
travis --> travis: update
review --> travis: update
travis --> review: wait_for_review
review --> coding: request_change
review --> merged: accept
review --> closed: reject
closed --> review: reopen
@enduml
```

As PlantUML let us define styles, I have provided some by default that the user can override.
Adding some marking:
```bash
php bin/console workflow:dump pull_request travis review --dump-format=puml
```
will give us:

```
@startuml
sprite $sf_logo [81x20/16z] {
hPNRaYiX24K1xwBo_tyx6-qaCtDEJ-KXLYMTLbp0HWcHZr3KRDJ8z94HG3jZn4_mijbQ2ryJoFePtXLWA_qxyGy19DpdY_10z11ZAbGjFHRwcEbcKx5-wqsV
yIMo8StMCHKh8ZUxnEwrZiwRAUOvy1lLcPQF4lEFAjhzMd5WOAqvKflS0Enx8PbihiSYXM8ClGVAseIWTAjCgVSAcnYbQG79xKFsZ0VnDCNc7AVBoPSMcTsX
UnrujbYjjz0NnsObkTgnmolqJD4QgGUYTQiNe8eIjtx4b6Vv8nPGpncn3NJ8Geo9W9VW2wGACm_JzgIO8A8KXr2jUBCVGEAAJSZ6JUlsNnmOzmIYti9G7bjL
8InaHM9G40NkwTG7OxrggvNIejA8AZuqyWjOzTIKi-wwYvjeHYesSWuPiTGDN5THzkYLU4MD5r2_0PDhG7LIUG33z5HtM6CP3icyWEVOS61sD_2ZsBfJdbVA
qM53XHDUwhY0TAwPug3OG9NonRFhO8ynF3I4unuAMDHmSrXH57V1RGvl9jafuZF9ZhqjWOEh98y0tUYGsUxkBSllIyBdT2oM5Fn2-ut-fzsq_cQNuL6Uvwqr
knh4RrvOKzxZfLV3s0rs_R_1SdYt3VxeQ1_y2_W2
}
title pull_request
skinparam titleBorderRoundCorner 15
skinparam titleBorderThickness 2
skinparam state {
    BackgroundColor<<initial>> #87b741
    BackgroundColor<<marked>> #3887C6
    BorderColor #3887C6
    BorderColor<<marked>> Black
    FontColor<<marked>> White
}
state start <<initial>>
state coding
state travis <<marked>>
state review <<marked>>
state merged
state closed
start --> travis: submit
coding --> travis: update
travis --> travis: update
review --> travis: update
travis --> review: wait_for_review
review --> coding: request_change
review --> merged: accept
review --> closed: reject
closed --> review: reopen
footer \nGenerated by <$sf_logo> **Workflow Component** and **PlantUML**
@enduml
```

Which gives you that:
![workflow](https://user-images.githubusercontent.com/313532/32086584-253e39c0-ba8b-11e7-82c7-fa24309dbcd6.png)

With `square` as transition, it gives you that:
![workflow](https://user-images.githubusercontent.com/313532/32533123-a8ea4530-c403-11e7-8f88-4f18d5e23a28.png)

Hope you will find that interesting!

Plopix committed:
Add option to the workflow:dump command to allow PlantUML format dump

Plopix commented:
That is failing on Windows, but I don't really know why, probably a weird Windows specificity, or the END OF LINE?
You might have an idea?

Thanks!

stloyd commented on src/Symfony/Component/Workflow/CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 CHANGELOG
 =========
 
+3.5.0
```
It will be `4.1.0` not `3.5.0` (which never will be released)

Plopix replied:
ok

stloyd commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        return
+            '@startuml'.PHP_EOL.static::SYMFONY_LOGO.PHP_EOL;
```
Should be in previous line.

Plopix replied:
ok

stloyd commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,153 @@
+
+    const MARKED = 'marked';
+
+    const SYMFONY_LOGO = 'sprite $sf_logo [81x20/16z] {
```
Better & probably safer than string comparison would be to save them as fixtures & compare results with pregenerated fixture.

Plopix replied:
I am not sure I understand here what you are proposing. Can you tell me more and I will be happy to change.

Plopix replied:
I think I got it and I made changes accordingly.

Plopix commented:
- I fixed the comment 9 (cc @stloyd)
- _appveyor_ build is passing now
- _travis-ci_ build fails because of VarDumper (that is not related to this PR cc @nicolas-grekas)

@Nyholm or @lyrixx
**Question/Note**: The current _workflow:dump_ does not generate the same way workflow and state machine. Is that a standard? I feel that it is so much more readable with transition names on the arrows. Is that something you want me to add on the PR? I am not a graph expert and it might be a convention or something.

nicolas-grekas commented:
DST time, transient failure will be fixed tomorrow hopefully :)

lyrixx commented:
> Question/Note: The current workflow:dump does not generate the same way workflow and state machine. Is that a standard?

Oups Sorry I forgot to reply.

Yes it's wanted. As a state machine can have many transition with the same name, the graph could be very ugly. that's why we did this.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -57,22 +61,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
             throw new \InvalidArgumentException(sprintf('No service found for "workflow.%1$s" nor "state_machine.%1$s".', $serviceId));
         }
 
+        /** @var \Symfony\Component\Workflow\Workflow $workflow */
```
Looks like it's useless

Plopix replied:
not for PHPStorm ;) but I can remove it

Plopix commented:
@lyrixx I have added an option to let the user decide which transition he wants.
I hope that is fine for you! Because to me, the square as transition kind of complexify sometimes too. It is great to have the option.

lyrixx commented:
Thanks. I will try your PR as soon as possible ;)

nicolas-grekas commented:
@lyrixx time to review? :)

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,199 @@
+
+    public function __construct($transitionType)
+    {
+        $this->transitionType = static::DEFAULT_TRANSITION_TYPE;
```
You don't need to do this since it's already defaulted when you do `$input->getOption('transition')`

Plopix replied:
I agree but this class could be used differently and without the command then I think it is correct to initialize it. no?

Simperfit replied:
Why not using 
```php
protected $transitionType = self::DEFAULT_TRANSITION_TYPE;
```
 instead ?
I may be missing something but it feels better to me.

Plopix replied:
No that is me, I don't know I wanted to avoid that because of php versions but now for 4.x it is php 7 then we are all good to do that. And I think we were good before too...I was wrong. It is better now ;)

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,199 @@
+    public function __construct($transitionType)
+    {
+        $this->transitionType = static::DEFAULT_TRANSITION_TYPE;
+        if ($transitionType == static::SQUARE_TRANSITION_TYPE) {
```
```php
if (static::SQUARE_TRANSITION_TYPE === $transitionType)
```

Plopix replied:
same as above comment

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,199 @@
+
+    protected function isSquareTransitionType()
+    {
+        return $this->transitionType == static::SQUARE_TRANSITION_TYPE;
```
why not `===` ?

Plopix replied:
ok will change

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,199 @@
+                $code[] = "agent {$transition->getName()}";
+            }
+        }
+        foreach ($definition->getTransitions() as $transition) {
```
It feels like alot of complexity comparing to the GraphvizDumper, isn't in the class something you could re-use to get arrays and then transform them directly ?

Plopix replied:
Not really sure I get what you want me to do. I agree it is not really straightforward but it is an efficient way to manage both TransitionType, GraphvizDumper as 2 classes to do that. I did not see the value for the PlantUML one, and I still don't see it. (but I would love to have your opinion)

I could do that in another method and then merge arrays... is it what you would like?.

Let me know!

Thank you!

Simperfit replied:
I was just thinking on how much this adds complexity. I guess if merging the arrays cost more in performance than what you are doing here let's keep this. (I think it will be anyway).

I just wanted to know more about this ;).

Plopix replied:
ok

Plopix commented:
I have updated the PR and rebased. Not sure why the tests are failing I think it is not related to this code, `master` code is failing too.

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+
+    public function __construct($transitionType = null)
+    {
+        if ($transitionType === static::SQUARE_TRANSITION_TYPE) {
```
You should use yoda style, btw I don't know if it's really needed to do this

Plopix replied:
- yoda style done.
- it is a way to kind of enforcing the 2 types and those 2 type only the default one or `square`
It would work without, but the value of `transitionType` could be inconsistent. Let me know your thoughts

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+
+    protected function isSquareTransitionType()
+    {
+        return $this->transitionType === static::SQUARE_TRANSITION_TYPE;
```
yoda style please.

Plopix replied:
done

Simperfit commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+        if (isset($options['name'])) {
+            $code[] = "title {$options['name']}";
+        }
+        if (isset($options['skinparams']) && count($options['skinparams']) > 0) {
```
I don't think you need to count here

Plopix replied:
correct I replace with is_array to ensure the foreach won't fail if someone send something else

Simperfit reviewed (approved):
Some minor comments, but great improvment thanks !

Plopix commented:
updated @Simperfit (we are almost there ;) )

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -19,6 +19,8 @@
 use Symfony\Component\Workflow\Dumper\GraphvizDumper;
 use Symfony\Component\Workflow\Dumper\StateMachineGraphvizDumper;
 use Symfony\Component\Workflow\Marking;
+use Symfony\Component\Console\Input\InputOption;
```
Input Option Is already imported.

Plopix replied:
fixed.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -19,6 +19,8 @@
 use Symfony\Component\Workflow\Dumper\StateMachineGraphvizDumper;
 use Symfony\Component\Workflow\Marking;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Workflow\Dumper\PlantUmlDumper;
```
Could you sort all `use`s?

Plopix replied:
ok

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +41,23 @@ protected function configure()
                 new InputArgument('marking', InputArgument::IS_ARRAY, 'A marking (a list of places)'),
                 new InputOption('label', 'l', InputArgument::OPTIONAL, 'Labels a graph'),
+                new InputOption('format', null, InputArgument::OPTIONAL, 'The dump format', 'dot'),
+                new InputOption(
```
We don't do that in Symfony. All arguments should be on one line..

Plopix replied:
alright, that is fixed.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +41,23 @@ protected function configure()
                 new InputOption('label', 'l', InputArgument::OPTIONAL, 'Labels a graph'),
+                new InputOption('format', null, InputArgument::OPTIONAL, 'The dump format', 'dot'),
+                new InputOption(
+                    'transition',
```
`puml_transition_format`

Plopix replied:
fixed.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +41,23 @@ protected function configure()
+                new InputOption(
+                    'transition',
+                    null,
+                    InputArgument::OPTIONAL,
```
This is wrong:
* It's an option, not an arguement
* The value is required

Plopix replied:
correct, fixed.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +41,23 @@ protected function configure()
 
-    %command.full_name% <workflow name> | dot -Tpng > workflow.png
+<info>DOT</info>:  %command.full_name% <workflow name> | dot -Tpng > workflow.png 
+<info>PUML</info>: %command.full_name% <workflow name> --format=puml -p | java -jar plantuml.jar > workflow.png
```
What is `-p` ?

Plopix replied:
That is a typo (not present in the PR doc) the `-p`, as mentioned in the PR doc is for `piping`. But it is for plantuml.jar... not for the command. I have fixed that too.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +41,23 @@ protected function configure()
                 new InputArgument('name', InputArgument::REQUIRED, 'A workflow name'),
                 new InputArgument('marking', InputArgument::IS_ARRAY, 'A marking (a list of places)'),
                 new InputOption('label', 'l', InputArgument::OPTIONAL, 'Labels a graph'),
+                new InputOption('format', null, InputArgument::OPTIONAL, 'The dump format', 'dot'),
```
"The dump format [dot|puml]"

lyrixx replied:
format is a bad name because it's already used in the --help command.
So `bin/console  workflow:dump article --format=puml -h ` do not work anymore

Plopix replied:
good catch I have renamed by `dump_format`

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -59,16 +71,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
+        if ('puml' === $format) {
+            $dumper = new PlantUmlDumper($transitionType);
+        } else {
```
use an else if here

Plopix replied:
ok done.

lyrixx reviewed (changes requested):
I let few comments on the Command Class. It seems you did not test it on real
project :/

And I followed you doc and it does not work. the program open a dialog windows
where I'm supposed to select a folder. I'm a bit lost.

Finally, I'm not sure if plantuml is really mainstream. By default the website
is in French :/ http://plantuml.com/

So for now I'm not sure we want this support in the Core. What do others think?

ping @xabbuh @Nyholm

Plopix commented:
Thank you for your review @lyrixx, I have implemented the changes, please find my feedbacks to your other comments.

> I let few comments on the Command Class. It seems you did not test it on real project :/

I don't know what you mean here, but this code was tested and was working. You caught good fixes that is really nice of you. I did not test on a *real* *real* project but I have:
- installed SF4 skeleton
- required the workflow component
- added a workflow configuration 
- added an entity class 
- tested the command (for real)

> And I followed you doc and it does not work. the program open a dialog windows where I'm supposed to select a folder. I'm a bit lost.

The doc in the PR mentions: "don't forget the -p to enable the "piping", but there was a typo in the doc in the command (`-p` was misplaced). Thanks for catching that!

> Finally, I'm not sure if plantuml is really mainstream. By default the website is in French :/ http://plantuml.com/

Well it is in English for me, are you sure it is not switched based on your browser?
Not sure if the language is relevant though. Look here it seems that it is used by many and for a long time:
https://en.wikipedia.org/wiki/PlantUML

And it does not invalidate the fact that it generates **nicer** diagrams, **more readable** and most of all **easier to adapt** diagram.

> So for now I'm not sure we want this support in the Core. What do others think?

That is your decision, feel free to deny, I was happy to do it anyway ;-). This PR is based on the comments I got at a lightning talk at the Symfony Live SF 2017. People found it interesting at this time. cc @nicolas-grekas

nicolas-grekas commented:
I still like it and would like to have it in core :)

lyrixx commented:
> I don't know what you mean here, but this code was tested and was working.

I said that because of the double use import. It made a fatal error. My comment was a bit harsh. I'm sorry about it.

---

About having it in the core, if @nicolas-grekas is OK, I'm also OK.
I'll review it another time today to ensure everything is OK. 

Thanks.

Plopix commented:
@lyrixx the duplicate was introduced when I rebased/merged. (because the new Label option that was not existing). The tests passed, I did not retest everything at this time.
No problem, comments are true and make SF solid. I don't mind ;)

cool to have it in the core, I am happy to contribute, and will be happy to do other changes if required!

lyrixx commented:
I just tested and now it works. Good job ;)

I will review the code, but could you remove the footer? (`Generated by Symfony .... `)

For everyone, I'm posting the workflow images of this [my demo app](https://github.com/lyrixx/SFLive-Paris2016-Workflow) (It's now powered by SF4 & Flex ;))

![workflow](https://user-images.githubusercontent.com/408368/34414424-9bc76f00-ebe9-11e7-9cd3-ce66d613eb2d.png)
![workflow2](https://user-images.githubusercontent.com/408368/34414425-9bdf61b4-ebe9-11e7-93d1-f0345a400e5b.png)

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +40,17 @@ protected function configure()
                 new InputArgument('name', InputArgument::REQUIRED, 'A workflow name'),
                 new InputArgument('marking', InputArgument::IS_ARRAY, 'A marking (a list of places)'),
                 new InputOption('label', 'l', InputArgument::OPTIONAL, 'Labels a graph'),
+                new InputOption('dump_format', null, InputOption::VALUE_REQUIRED, 'The dump format [dot|puml]', 'dot'),
```
I know I tell you to use `dump_format` but actually It should  be `dump-format` (dash instead of underscore)
And It's the same for `puml_transition_format`

Sorry :/

Plopix replied:
done.

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -39,13 +40,17 @@ protected function configure()
-    %command.full_name% <workflow name> | dot -Tpng > workflow.png
+<info>DOT</info>:  %command.full_name% <workflow name> | dot -Tpng > workflow.png 
+<info>PUML</info>: %command.full_name% <workflow name> --dump_format=puml | java -jar plantuml.jar -p > workflow.png
+<info>PUML</info>: %command.full_name% <workflow name> --dump_format=puml --puml_transition_format=square | java -jar plantuml.jar -p > workflow.png
```
Don't forget to update this too

Plopix replied:
ok

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -59,16 +64,27 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $container = $this->getApplication()->getKernel()->getContainer();
         $serviceId = $input->getArgument('name');
+        $format = $input->getOption('dump_format');
+        $transitionType = $input->getOption('puml_transition_format');
```
This var declaration could be (re)moved to line 81

Plopix replied:
moved

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -59,16 +64,27 @@ protected function execute(InputInterface $input, OutputInterface $output)
     {
         $container = $this->getApplication()->getKernel()->getContainer();
         $serviceId = $input->getArgument('name');
+        $format = $input->getOption('dump_format');
```
I like to declare var very near where I used them. Could you move it line 80?

Plopix replied:
moved

lyrixx commented on src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php:
```diff
@@ -80,6 +96,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
         if (null !== $label && '' !== trim($label)) {
             $options = array('graph' => array('label' => $label));
         }
+        $options += array('name' => $serviceId);
```
Could you use array_replace? (more "symfony style")

Plopix replied:
done

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+}';
+
+    const DEFAULT_TRANSITION_TYPE = 'arrow';
+
```
all blank lines to be removed (only between all constants)

Plopix replied:
done

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+
+    const SQUARE_TRANSITION_TYPE = 'square';
+
+    protected static $defaultOptions = array(
```
the default PHP version is now 7+. So it could be a constant

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+        ),
+    );
+
+    protected $transitionType = self::DEFAULT_TRANSITION_TYPE;
```
move everything protected to private

Plopix replied:
I still don't understand that. I understand the concept of not allowing inheritance first. But for a dumper, I don't see why it would be bad to extend it and tweak stuff in a subclass.
I would love a point of view.

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+
+    protected $transitionType = self::DEFAULT_TRANSITION_TYPE;
+
+    public function __construct($transitionType = null)
```
could be typehinted string.

Plopix replied:
done

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+
+    public function __construct($transitionType = null)
+    {
+        if (static::SQUARE_TRANSITION_TYPE === $transitionType) {
```
if (!array($transitionType, [...]) throw new InvalidArgumentException

$this->transitionType = $transitionType;

Plopix replied:
done

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+    /**
+     * @param Definition   $definition
+     * @param Marking|null $marking
+     * @param array        $options
```
This PHP Doc should be removed

Plopix replied:
done

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    protected function isSquareTransitionType()
```
This method should be private, and after public ones

Plopix replied:
same comment about private over protected ;)

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+            $code[] =
+                "state {$place}".
+                ($definition->getInitialPlace() === $place ? ' <<'.self::INITIAL.'>>' : '').
+                ($marking && $marking->has($place) ? ' <<'.self::MARKED.'>>' : '');
```
@nicolas-grekas Are you OK with this CS ?

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+    /**
+     * @param array $options
+     *
+     * @return string
```
PHP doc be removed (like others).
You can add the return type hint :)

(actually, PHPDoc do not add value. That's why I'm asking to remove it)

Plopix replied:
done

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+     */
+    protected function startPuml(array $options)
+    {
+        $start = '@startuml'.PHP_EOL;
```
I don't know how the .jar works, but I'm pretty sure we could **always** use "\n" here. It will be simpler.

Plopix replied:
simpler to use "\n" over `PHP_EOL`? why?

lyrixx commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,198 @@
+            $start .= 'allow_mixing'.PHP_EOL;
+        }
+
+        if (isset($options['nofooter']) && true === $options['nofooter']) {
```
PHP7 FTW: (if $options['nofooter'] ?? false) )
But anyway, let's remove this default footer

Plopix replied:
done, but I kept the footer option. It is just not by default anymore in the command.
Ok with that?

lyrixx commented on src/Symfony/Component/Workflow/Tests/Dumper/PlantUmlDumperTest.php:
```diff
@@ -0,0 +1,102 @@
+use Symfony\Component\Workflow\Tests\WorkflowBuilderTrait;
+
+/**
+ * Class PlantUmlDumperTest.
```
To be removed.
And IIRC the rule is to not add `@author` on Test class

Plopix replied:
ok done.

lyrixx reviewed (changes requested):
I let some easy comments.

Thanks for your work. It's almost finish. \o/

Plopix commented:
thanks @lyrixx, I have done most of the comments and asked questions about some of them.
Let me know, I will finalize ASAP.

Plopix commented:
@lyrixx are we good here? thx!

lyrixx commented:
:+1: (I will merge it this week if everyone is OK :tada: )

fabpot commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,171 @@
+/**
+ * PlantUmlDumper dumps a workflow as a PlantUML file.
+ *
+ * You can convert the generated dot file with the plantuml.jar utility (http://plantuml.com/):
```
I suppose it does not generate a dot file.

Plopix replied:
that is correct, I have pushed the correction. thx!

stof commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,171 @@
+knh4RrvOKzxZfLV3s0rs_R_1SdYt3VxeQ1_y2_W2
+}';
+    const INITIAL = 'initial';
+    const MARKED = 'marked';
```
shouldn't some of these be private constants ?

stof commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,171 @@
+
+    private function isSquareTransitionType(): bool
+    {
+        return static::SQUARE_TRANSITION_TYPE === $this->transitionType;
```
use `self::` to access constant. There is no reason to make them an extension point by allowing changing their values (and the code would not work anyway in such case, due to other places)

stof commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,171 @@
+            return $start;
+        }
+
+        return $start.static::SYMFONY_LOGO.PHP_EOL;
```
`self:` here as well

stof commented on src/Symfony/Component/Workflow/Dumper/PlantUmlDumper.php:
```diff
@@ -0,0 +1,171 @@
+            return $end;
+        }
+
+        return PHP_EOL.'footer \nGenerated by <$sf_logo> **Workflow Component** and **PlantUML**'.$end;
```
Are you sure about the `\n` here, which is not a newline char ?

Plopix replied:
yes I think so, the `\n` is working as expected.

stof commented:
The dump of the workflow in https://github.com/symfony/symfony/pull/24705#issuecomment-354303061 looks confusing to me. It looks like there are 2 independant `publish` transitions, allowing to go one path or the other, while it is actually a single transition with 2 inputs (meaning that it is a AND between both paths, not a OR). This is exactly the reason why the dot dumper uses boxes to represent transitions, not arrows (to allow having multiple inputs and multiple outputs for each of them).
This is the main difference between a generic workflow and a state machine (the state machine has only one input and one output for each transition)

lyrixx commented:
Oh @stof Is totally right. This PR could not be merged in this state.

stof commented:
The current dumper is fine for the case of state machines though

Plopix commented:
@stof thank you I have updated the code.
About the confusion, well as I said before, I am not an expert on workflow/state_machine, I was just trying to provide a better dump.

From my point of view, I don't find it that confusing as I know that I can have only one `transition` with the same name, then it must be the same.

You said:
> This is the main difference between a generic workflow and a state machine (the state machine has only one input and one output for each transition)
But here: https://symfony.com/doc/current/workflow/state-machines.html
`update` gets multiple 'from'


The PR is aiming to let the choice to the user to dump it with `arrow` or `square` transition.
I have compared closely the dumps with Dot and PlantUML
On the case provided in the doc and on the `article` workflow of @lyrixx :

Pull Request: https://symfony.com/doc/current/workflow/state-machines.html
<img width="1650" alt="screen shot 2018-01-23 at 1 46 31 pm" src="https://user-images.githubusercontent.com/313532/35303335-7400acea-0046-11e8-8e02-e51752ce6c96.png">

Blog Publishing: https://symfony.com/doc/current/workflow/usage.html
<img width="1655" alt="screen shot 2018-01-23 at 1 49 20 pm" src="https://user-images.githubusercontent.com/313532/35303338-7588c4f8-0046-11e8-8e9b-4924d42a4ae7.png">

Article: https://github.com/lyrixx/SFLive-Paris2016-Workflow/blob/master/config/packages/workflow.yaml#L3
<img width="1606" alt="screen shot 2018-01-23 at 1 51 03 pm" src="https://user-images.githubusercontent.com/313532/35303340-77a705d8-0046-11e8-91ae-0dd4f7d206d7.png">

I think that is working, I am not sure what you are expecting, I see at least 3 options in which we could go:

- 1/ we could force `square` for `workflow` and `arrow` for `state_machine`
- 2/ we could put by default `square` for `workflow` and `arrow` for `state_machine`
- 3/ don't do change anything as

Let me know

lyrixx commented:
>1/ we could force square for workflow and arrow for state_machine

It's definitively the way to go ;)

Plopix commented:
hey @lyrixx @stof I have done the option 1/

For the sake of my understanding, I would love if you could you explain why "arrows" are not valid sometimes?
<img width="1087" alt="screen shot 2018-01-25 at 12 07 02 am" src="https://user-images.githubusercontent.com/313532/35377512-534d79e6-0164-11e8-8043-3f67bb003072.png">

In this example, for instance, it feels simpler to have arrows. But I would understand if the reason is to force a kind of consistency for each type.

Let me know!

lyrixx commented:
![2018-01-26 10 38 57](https://user-images.githubusercontent.com/408368/35433901-5d22f6ea-0285-11e8-8e91-889272802094.jpg)

As you can see on the previous schema, If you don't draw the transition (square), you have **no way** to know how many transition there are. And of course the 2 workflows (line 1 and line 2) are totally different.

Plopix commented:
ok makes sense ;-) all good to merge then maybe ;-)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

fabpot commented:
Thank you @Plopix.

lyrixx commented:
:clap: Thanks for your hard work on this feature. I just tested it and it works very well.
